recsys/recsys_tutorial.py

```diff
@@ -0,0 +1,286 @@
+
+book_to_first_author = dict(zip(df_books.book_idx, df_books.first_author))
+first_author_to_books_df = df_books.groupby("first_author")[["book_idx"]].agg(set)
+first_author_to_books = dict(
+    zip(first_author_to_books_df.index, first_author_to_books_df.book_idx)
+)
+
+
+@labeling_function()
+def common_first_author(x):
+    author = book_to_first_author[x.book_idx]
+    same_author_books = first_author_to_books[author]
+    num_read = len(set(x.book_idxs).intersection(same_author_books))
+    return 1 if num_read > 8 else -1
+
+
+# %% [markdown]
+# We can also leverage the long text reviews written by users to guess whether they liked or disliked a book. We run [TextBlob](https://textblob.readthedocs.io/en/dev/index.html), a tool that provides a pretrained sentiment analyzer, on the reviews, and use its polarity and subjectivity scores to estimate the user's rating for the book.
+
+# %%
+from snorkel.preprocess import preprocessor
+from textblob import TextBlob
+
+
+@preprocessor()
+def textblob_polarity(x):
+    if x.review_text:
+        x.blob = TextBlob(str(x.review_text))
+        x.start_blob = TextBlob(" ".join(x.blob.raw_sentences[:2]))
+    else:
+        x.blob = None
+    return x
+
+
+textblob_polarity.memoize = True
+
+# Label high polarity reviews as positive.
+@labeling_function(pre=[textblob_polarity])
+def polarity_positive(x):
+    if x.blob:
+        if x.blob.polarity > 0.3:
+            return 1
+    return -1
+
+
+# Label high subjectivity reviews as positive.
+@labeling_function(pre=[textblob_polarity])
+def subjectivity_positive(x):
+    if x.blob:
+        if x.blob.subjectivity > 0.75:
+            return 1
+    return -1
+
+
+# Label low polarity reviews as negative.
+@labeling_function(pre=[textblob_polarity])
+def polarity_negative(x):
+    if x.blob:
+        if x.blob.polarity < 0.0:
+            return 0
+    return -1
+
+
+# %%
+from snorkel.labeling import PandasLFApplier, LFAnalysis
+
+lfs = [common_first_author, polarity_positive, subjectivity_positive, polarity_negative]
+
+applier = PandasLFApplier(lfs)
+L_dev = applier.apply(data_dev)
+LFAnalysis(L_dev, lfs).lf_summary(data_dev.rating)
+
+# %% [markdown]
+# ### Applying labeling functions to the training set
+#
+# We apply the labeling functions to the training set, and then filter out examples unlabeled by any LF, and combine the rest with the dev set to form our final training set.
+
+# %%
+from snorkel.labeling.model.label_model import LabelModel
+
+# Train LabelModel.
+L_train = applier.apply(data_train)
+label_model = LabelModel(cardinality=2, verbose=True)
+label_model.fit(L_train, n_epochs=5000, seed=123, log_freq=20, lr=0.01)
+
+from snorkel.analysis import metric_score
+from snorkel.utils import probs_to_preds
+
+Y_dev_prob = label_model.predict_proba(L_dev)
+Y_dev_pred = probs_to_preds(Y_dev_prob)
+
+acc = metric_score(data_dev.rating, Y_dev_pred, probs=None, metric="accuracy")
+print(f"LabelModel Accuracy: {acc:.3f}")
```

Comment:
I actually got rid of the LabelModel accuracy altogether. It is misleading because our LFs have very low coverage (together they cover maybe 5% of the dataset). That is still enough training data for the model because the train set is so huge, but the LabelModel metrics are dominated by it's values on the uncovered examples.